gpu/preconditioner/block_jacobi_kernels.cu

```diff
@@ -166,6 +166,151 @@ GKO_ENABLE_IMPLEMENTATION_SELECTION(select_apply, apply);
 using compiled_kernels = syn::compile_int_list<1, 13, 16, 32>;
 
 
+template <typename IndexType>
+__device__ __forceinline__ bool has_same_nonzero_pattern(
+    const IndexType *prev_row_ptr, const IndexType *curr_row_ptr,
+    const IndexType *next_row_ptr)
+{
+    if (next_row_ptr - curr_row_ptr != curr_row_ptr - prev_row_ptr) {
+        return false;
+    }
+    for (; curr_row_ptr < next_row_ptr; ++prev_row_ptr, ++curr_row_ptr) {
+        if (*curr_row_ptr != *prev_row_ptr) {
+            return false;
+        }
+    }
+    return true;
+}
+
+template <typename IndexType>
+__global__ void find_natural_blocks_kernel(
+    size_type num_rows, int32 max_block_size, const IndexType *row_ptrs,
+    const IndexType *col_idx, IndexType *block_ptrs, size_type *num_blocks_arr)
+{
+    block_ptrs[0] = 0;
+    if (num_rows == 0) {
+        return;
+    }
+    size_type num_blocks = 1;
+    int32 current_block_size = 1;
+    for (size_type i = 1; i < num_rows; ++i) {
+        const auto prev_row_ptr = col_idx + row_ptrs[i - 1];
+        const auto curr_row_ptr = col_idx + row_ptrs[i];
+        const auto next_row_ptr = col_idx + row_ptrs[i + 1];
+        if (current_block_size < max_block_size &&
+            has_same_nonzero_pattern(prev_row_ptr, curr_row_ptr,
+                                     next_row_ptr)) {
+            ++current_block_size;
+        } else {
+            block_ptrs[num_blocks] =
+                block_ptrs[num_blocks - 1] + current_block_size;
+            ++num_blocks;
+            current_block_size = 1;
+        }
+    }
+    block_ptrs[num_blocks] = block_ptrs[num_blocks - 1] + current_block_size;
+    num_blocks_arr[0] = num_blocks;
+}
+
+
+template <typename ValueType, typename IndexType>
+size_type find_natural_blocks(std::shared_ptr<const GpuExecutor> exec,
+                              const matrix::Csr<ValueType, IndexType> *mtx,
+                              int32 max_block_size, IndexType *block_ptrs)
+{
+    auto rows = mtx->get_num_rows();
+    auto cols = mtx->get_num_cols();
+    auto vals = mtx->get_const_values();
+    auto row_ptrs = mtx->get_const_row_ptrs();
+    auto col_idx = mtx->get_const_col_idxs();
+
+    Array<size_type> d_nums_array(exec, 1);
+    auto d_nums = d_nums_array.get_data();
+
+    Array<size_type> nums_array(exec->get_master(), 1);
+    auto nums = nums_array.get_data();
+
+    constexpr int64 blocksize3_1 = 1;
+    constexpr int64 blocksize3_2 = 1;
+    constexpr int64 blocksize3_3 = 1;
+
+    const int64 dimgrid3_1 = 1;
+    const int64 dimgrid3_2 = 1;
+    const int64 dimgrid3_3 = 1;
+
+    dim3 grid3(dimgrid3_1, dimgrid3_2, dimgrid3_3);
+    dim3 block3(blocksize3_1, blocksize3_2, blocksize3_3);
+
+    find_natural_blocks_kernel<<<grid3, block3, 0, 0>>>(
```

Comment:
I do agree that block-diagonal detection has an O(num_rows) sequential portion which is hard to parallelize.
However,  I do not agree that it has an O(num_nonzeros) sequential portion, as is the case in this kernel.

Why not parallelize the comparison of adjacent rows? (I believe this has already been done in MAGMA-sparse?) You could also fix the problem of non-coalesced accesses to indexes by assigning a (sub-)warp to work on a single row comparison. The same idea as in the batched CSR SpMV kernel (the unbalanced version) we have in MAGMA-sparse.

For the sequential part, I am really not sure we want to do it on the GPU - a single GPU core is __extremely__ slow. 
According to the results from the balanced SpMV paper, just the imbalance caused a 100x slowdown of the cuSPARSE CSR matrix-vector product over the balanced version (which is by itself not a very "fast" kernel, as it does tons of extra operations). We can view sequential execution as an extreme case of imbalance - which would suggest that just traversing the matrix sequentially on the GPU could be more than 100x slower than a single SpMV.

Reply:
I am not sure this has O(num_nonzeros) complexity. It only has O(num_nonzeros) complexity if all rows have the same number of nonzero elements. So the complexity ranges from O(num_rows) to O(num_nonzeros). 
In MAGMA-sparse I have implemented it using multiple kernels: first comparing the number of nonzeros in different rows. Then, a second kernel comparing the elements in the rows that have the same number of nonzero elements. This adds 2n memory access (writing out which rows have the same nz_count, reading that information in the second kernel). Maybe it is possible doing better using dynamic parallelism, but maybe this is over-engineering.
For aligned access, the second kernel  should use thread blocks of size >= nz_count in this row. But can we assume that no row has more than 1024 elements?

Reply:
> I am not sure this has O(num_nonzeros) complexity. It only has O(num_nonzeros) complexity if all rows have the same number of nonzero elements. So the complexity ranges from O(num_rows) to O(num_nonzeros).

You could use the same reasoning to claim that the complexity of binary search ranges between O(1) and O(log n) - technically true, but everyone will talk about worst-case complexity unless they state otherwise - the same as I did here. There is also the concept of "average-case" complexity, like in quick-sort which is usually considered O(n log n) algorithm even though it's worst case complexity is O(n^2), since it has been proven that this is the usual time it takes.

I won't try to formally analize the average-case complexity of this thing, but there are some points which can intuitively help us determine that it is not O(num_rows). First of all, there are 2 possible outcomes of the blocking procedure:

1.  A positive outcome: the algorithm finds a block structure (let's intuitively say this is the case if at least half of the rows belong to a block - let's denote the number of such rows with _B_)
2.  A negative outcome: the algorithm does not find a block structure (intuitively, negation of 1.)

If we define _K_ := "the number of positions _i_ s.t. rows _i_ and _i_+1 have the same nonzero pattern", then we get a simple bound for _B_:  _K_ &le; _B_ &le; 2 _K_, or equivalently for _K_: _B_ / 2 &le; _K_ &le; _B_.
Assuming a relatively balanced nonzero pattern, the number of column indexes of the matrix we will have to traverse is going to be _K_ * (_num_nonzeros_ / _num_rows_).

In case we have a positive outcome, we know that _K_ &ge; _B_ / 2 &ge; _num_rows_ / 4, and thus the number of nonzeros we have to traverse will be at least (_num_rows_ / 4) * (_num_nonzeros_ / _num_rows_) = _num_nonzeros_ / 4. Thus, for any input that results in a positive outcome the complexity will be O(_num_nonzeros_).

I believe a reasonable assumption is that the user knows something about their matrix, and they won't often try to even call the algorithm on a matrix that doesn't have a block structure, thus in most cases the result will be positive, and hence the average complexity will be O(_num_nonzeros_).

> In MAGMA-sparse I have implemented it using multiple kernels: first comparing the number of nonzeros in different rows. Then, a second kernel comparing the elements in the rows that have the same number of nonzero elements. This adds 2n memory access (writing out which rows have the same nz_count, reading that information in the second kernel). 

Firs of all, that's 3 * _n_, not 2 * _n_ memory accesses, and why not just merge the two into a single kernel that first checks if the rows have the same length, and if they do, compares the nonzero locations?
In any case, a single sequential computation on a GPU is about 10^4 times slower than the memory bandwidth  (0.01 GFLOP/s sequential compute vs ~125 GFloat/s bandwidth on a P100, see my axpy tests on lecture 5 of your HPC course @ KIT), so replacing a parallel code with a sequential one just to reduce the amount of memory transferred is usually not a good idea.

> Maybe it is possible doing better using dynamic parallelism, but maybe this is over-engineering.

I don't see how dynamic parallelism would help here.

> For aligned access, the second kernel should use thread blocks of size >= nz_count in this row. 

Not sure I understand this. Since the cache line is 128 bytes (equivalently, 32 integers) long, every access that starts on and address that is a multiple of this value is aligned. Sure, we can do something smarter to ensure aligned accesses to everything, but I don't think that gives us a lot of performance. Just ensuring that we have coalesced access (which we can do by assigning an entire warp per row) should give us a lot better performance than sequential code.

> But can we assume that no row has more than 1024 elements?

Obviously no, but I don't get why we would need this assumption.

If it wasn't clear what kind of parallelization strategy I'm proposing, it's the same idea as in [vector CSR SpMV by BG](http://www.nvidia.com/docs/IO/77944/sc09-spmv-throughput.pdf).